wallet: Various fixes and cleanup to keypool handling in LegacyScriptPubKeyMan and CWallet

achow101 opened the pull request:
* The `pwallet->CanGetAddresses()` call in `ReserveDestination::GetReservedDestination` to `LegacyScriptPubKeyMan::GetReservedDestination` so that the sanity check results in a failure when a `ScriptPubKeyMan` individually cannot get a destination, not when any of the `ScriptPubKeyMan`s can't.
* `ScriptPubKeyMan::GetReservedDestination` is changed to return the destination so that future `ScriptPubKeyMan`s can return destinations constructed in other ways. This is implemented for `LegacyScriptPubKeyMan` by moving key-to-destination code from `CWallet` to `LegacyScriptPubKeyMan`
* In order for `ScriptPubKeyMan` to be generic and work with future `ScriptPubKeyMan`s, `ScriptPubKeyMan::ReturnDestination` is changed to take a `CTxDestination` instead of a `CPubKey`. Since `LegacyScriptPubKeyMan` still deals with keys internally, a new map `m_reserved_key_to_index` is added in order to track the keypool indexes that have been reserved.
* A bug is fixed in how the total keypool size is calculated as it was omitting `set_pre_split_keypool` which is a bug.

Split from #17261

achow101 committed:
Key pool: Change ReturnDestination interface to take address instead of key

In order for ScriptPubKeyMan to be generic and work with future
ScriptPubKeyMans, ScriptPubKeyMan::ReturnDestination is changed to
take a CTxDestination instead of a CPubKey. Since LegacyScriptPubKeyMan
still deals with keys internally, a new map m_reserved_key_to_index is
added in order to track the keypool indexes that have been reserved.

The CPubKey argument of KeepDestination is also  removed so that it is
more generic. Instead of taking a CPubKey or a CTxDestination, we just use
the nIndex given to find the pubkey.

achow101 committed:
Key pool: Move CanGetAddresses call

Call LegacyScriptPubKeyMan::CanGetAddresses directly instead of calling
CWallet::CanGetAddresses to only query the relevant key manager

This is a minor change in behavior: call now only happens if a new key needs to
be reserved, since if a key is already reserved it might fail unnecessarily.

This change also serves as a sanity check
https://github.com/bitcoin/bitcoin/pull/16341#discussion_r331238394

achow101 committed:
Key pool: Fix omitted pre-split count in GetKeyPoolSize

This is a bugfix: https://github.com/bitcoin/bitcoin/pull/16341#discussion_r330669214

achow101 committed:
Key pool: Move LearnRelated and GetDestination calls

Addresses are determined by LegacyScriptPubKeyMan::GetReservedDestination
instead of ReserveDestination::GetReservedDestination as other ScriptPubKeyMan
implementations may construct addresses differently

This does not change behavior.

Sjors reviewed (commented):
In commit 79c0a7c `Key pool: Move TopUp call` @ryanofsky wrote:
> but may be "annoying" and lead to unnecessary topups in some situations: https://github.com/bitcoin/bitcoin/pull/16341#issuecomment-510903836

My comment there was:

> In the context of a hardware wallet I want `TopUp` to fetch keys from the device. But it should only try that if the user explicitly calls `topupkeypool` (and the UI equivalent). Problem is that TopUp doesn't know where it's called from.

I'm not worried about this anymore. Initially we're probably only supporting standard BIP44/49/84 style address derivation used by most hardware manufactures default wallets. In a descriptor wallet you always have the account `xpub`, so you can always topup the keypool without pinging the device.

In order to support hardened derivation we'd have to fetch individual public keys / destinations from the device, using a topup call. However in that case we can simply ignore a failed topup, unless we actually need the key. This is the direction e.g. #17219 is taking.

Re b1ba8b6e5ef3fe86f650c0df1042d539d455ef29 `Key pool: Change ReturnDestination interface to take address instead of key`.

The change seems useful, but incomplete. `CTxDestination` is more generic than `CPubKey`, which offers more flexibility with (e.g. multisig) descriptor wallets. But then I think we should use `CTxDestination` everywhere, so maybe:
* turn `std::map<int64_t, CKeyID> m_reserved_key_to_index` into `std::map<int64_t, CTxDestination > m_reserved_destination_to_index`; and/or
* `std::map<CKeyID, int64_t> m_pool_key_to_index;` to `std::map< CTxDestination, int64_t> m_pool_destination_to_index`

Maybe that's too many changes, but it might make things more readable. 

Also, @ryanofsky wrote:

> The address argument is currently unused, and not having the key argument
requires the legacy keyman to maintain a new `m_reserved_key_to_index` map, so
it's not clear what benefit of this change is, but it might help with hardware
wallet support.

I'm not sure if it does. Maintaining a map of indexes might be useful if we want to do fancy things like having two change address to throw of chain analytics.

Re d7affb5fc `Key pool: Make TopUp fail if unexpected wallet flags are set`. Currently in the descriptor wallet PR you have to specify a range when importing a descriptor, which determines what goes into the keypool. That's probably why the check here for `WALLET_FLAG_DISABLE_PRIVATE_KEYS` doesn't cause any tests to fail. But this `range` argument shouldn't be necessary for descriptors without hardened derivation after the `xpub` / `xpriv`. For those types of descriptors, a watch-only descriptor wallet has the same "keypool" (expanding the descriptor) functionality as a wallet with private keys. The check for `WALLET_FLAG_BLANK_WALLET` does make sense.

achow101 commented:
> The change seems useful, but incomplete. CTxDestination is more generic than CPubKey, which offers more flexibility with (e.g. multisig) descriptor wallets. But then I think we should use CTxDestination everywhere, so maybe:

I would prefer to clean this up in a followup.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -3242,9 +3242,6 @@ bool ReserveDestination::GetReservedDestination(const OutputType type, CTxDestin
 
-    if (!pwallet->CanGetAddresses(internal)) {
-        return false;
-    }
```
Does it really matter? It's removed a few commits later anyways.

Sjors replied:
No, it's a nit :-)

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -3246,16 +3246,14 @@ bool ReserveDestination::GetReservedDestination(const OutputType type, CTxDestin
         fInternal = keypool.fInternal;
+        address = dest;
     }
     assert(vchPubKey.IsValid());
```
Ignoring that it's removed later, the check matters here. since the pubkey should always be valid after GetReservedDestination. Either it was fetched earlier, or it was fetched just then.

ryanofsky commented:
I would love for somebody to do a novelization of this PR like I did for https://github.com/bitcoin/bitcoin/pull/17369#pullrequestreview-313636620. I think I understand what the changes here are doing, but not why they are being made. Particularly I don't understand motivation for commits:

- 8b9a1ed9fb5ef64385d3881b1423059e4daaef13 Key pool: Move TopUp call (3/6)
- 8fc748e8050525a4235e34f3d93f3849245177f0 Key pool: Change ReturnDestination interface to take address instead of key (4/6)
- 51ed5b0d73fc1e7010fc4b1bc26b4e1946cf60b0 Key pool: Make TopUp fail if unexpected wallet flags are set (5/6)

Also unclear if Sjors comment above about "Make TopUp fail" https://github.com/bitcoin/bitcoin/pull/17373#pullrequestreview-311826649 needs to be addressed

achow101 commented:
@ryanofsky I've updated the OP with something like that.

Sjors commented:
Thanks for the novelization.

> Also unclear if Sjors comment above about "Make TopUp fail" [#17373 (review)](https://github.com/bitcoin/bitcoin/pull/17373#pullrequestreview-311826649) needs to be addressed

I.e. `WALLET_FLAG_DISABLE_PRIVATE_KEYS` descriptor should be able to up their "keypool". Not too hard to change that later though.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -3252,9 +3252,6 @@ bool ReserveDestination::GetReservedDestination(const OutputType type, CTxDestin
         return false;
     }
 
-    if (!pwallet->CanGetAddresses(internal)) {
```
In commit "Key pool: Move CanGetAddresses call" (f034e5748aa2fac9fb8c9010c806d00b2b8d79ba)

Maybe say something like "Call LegacyScriptPubKeyMan::CanGetAddresses directly instead of calling CWallet::CanGetAddresses to only query the relevant key manager" in the commit description be clearer about what the change is supposed to do.

In case it helps anybody else, after 9b161a718f3c9d301419940cc9456f2065f8ca58 from #17261 there can more than keymanager and CWallet::CanGetAddresses will return false any of them can't generate addresses, so it wouldn't be right to call that method here.

achow101 replied:
Done

ryanofsky commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -271,6 +271,8 @@ bool LegacyScriptPubKeyMan::GetReservedDestination(const OutputType type, bool i
         return false;
     }
+    LearnRelatedScripts(keypool.vchPubKey, type);
+    address = GetDestinationForKey(keypool.vchPubKey, type);
```
In commit "Key pool: Move LearnRelated and GetDestination calls" (35c37b59c5bb918c15aca3092b6506ef3974e50c)

Maybe point out in the commit description that addresses are determined by LegacyScriptPubKeyMan::GetReservedDestination 
instead of ReserveDestination::GetReservedDestination because other key manager implementations might construct them in other ways.

achow101 replied:
Done

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -3076,8 +3076,6 @@ bool CWallet::GetNewChangeDestination(const OutputType type, CTxDestination& des
 {
     error.clear();
 
-    m_spk_man->TopUp();
```
In commit "Key pool: Move TopUp call" (8b9a1ed9fb5ef64385d3881b1423059e4daaef13)

Sorry if this is a dumb question, because my understanding here is not great, but the PR description says this call has to be moved so "it will work after future changes where m_spk_man is not a global thing in CWallet and is fetched based on type and internal-ness".

But why couldn't the code here just look up the right keyman object using the type above and the internal-ness which it already knows?

Not saying this shouldn't be moved, just that this doesn't seem to explain the move

achow101 replied:
The place it is moved to does the lookup as well. It was redundant to do the lookup twice.

achow101 replied:
Updated the commit message.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -3252,6 +3250,7 @@ bool ReserveDestination::GetReservedDestination(const OutputType type, CTxDestin
         return false;
     }
 
+    m_spk_man->TopUp();
```
In commit "Key pool: Move TopUp call" (8b9a1ed9fb5ef64385d3881b1423059e4daaef13)

Since aside from the call in `GetNewChangeDestination` this code is also called from `CWallet::CreateTransaction`, I'm wondering if this is a change in behavior. Is topup already called when creating transactions? Wondering if this the "unsollicited keypool topup" refered to by https://github.com/bitcoin/bitcoin/pull/16341#issuecomment-510903836

achow101 replied:
This is not a change in behavior because `TopUp` is called in `LegacyScriptPubKeyMan::ReserveKeyFromKeyPool`

achow101 replied:
> So if it's already called then why do we need to add a call here?

It's a belt-and-suspenders as not all ScriptPubKeyMans are guaranteed to do this.

I've moved it inside the block.

ryanofsky reviewed (commented):
Started review (will update list below with progress).
- [X] f034e5748aa2fac9fb8c9010c806d00b2b8d79ba Key pool: Move CanGetAddresses call (1/6)
- [X] 35c37b59c5bb918c15aca3092b6506ef3974e50c Key pool: Move LearnRelated and GetDestination calls (2/6)
- [X] 8b9a1ed9fb5ef64385d3881b1423059e4daaef13 Key pool: Move TopUp call (3/6)
- [ ] 8fc748e8050525a4235e34f3d93f3849245177f0 Key pool: Change ReturnDestination interface to take address instead of key (4/6)
- [ ] 51ed5b0d73fc1e7010fc4b1bc26b4e1946cf60b0 Key pool: Make TopUp fail if unexpected wallet flags are set (5/6)
- [ ] 228a316e6390e512668095331417a5ae25130170 Key pool: Fix omitted pre-split count in GetKeyPoolSize (6/6)

Sjors commented:
ACK eb9c2601377d40c4d55798a5ffcbe8c4791b91e5

Nits:
* you're not really selling 06d411488c1c669f00a1c947416267941dbb7f7b with the current commit message; maybe replace with what you wrote in the PR description.
* maybe call `reserved_key_to_index.erase(index)` in `KeepDestination()` as well for completeness
* `m_reserved_key_to_index` could be more accurately called `m_index_to_reserved_key` or even just `m_reserved_keys` (which begs the question why the keypool uses sets instead of maps, but that's for a later cleanup)

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -3049,6 +3049,9 @@ unsigned int CWallet::GetKeyPoolSize() const
 
 bool CWallet::TopUpKeyPool(unsigned int kpSize)
 {
+    if (IsWalletFlagSet(WALLET_FLAG_BLANK_WALLET) || IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
```
In commit "Key pool: Make TopUp fail if unexpected wallet flags are set" (b12121d52f46ac7e8455479bc9ded100cfccc841)

There should be a code comment here to explain what's mentioned in the PR description: that the blank wallet check is needed for TopUp to return false instead of true in a wallet without any keyman objects. It would also be good to say why that's important, because on the surface it seems like a bad choice to return failure when nothing failed.

Relatedly, I would love to be able to drop this commit if it isn't actually necessary or if there isn't a specific example of a bug you could imagine this check preventing. Wallet flag code is confusing and just referencing flags in random places for no apparent reason and with no explanation isn't the greatest thing.

It's one thing to add sanity checks in asserts or CHECK_NONFATAL calls, but something else to add sanity checks in real code without clear "this condition is never expected to happen" comments so people reading the code don't waste time trying to imagine cases where it could actually do anything.

achow101 replied:
Looking at this change further, I don't think it's needed, so I've removed this commit.

ryanofsky commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1185,6 +1186,7 @@ bool LegacyScriptPubKeyMan::ReserveKeyFromKeyPool(int64_t& nIndex, CKeyPool& key
             throw std::runtime_error(std::string(__func__) + ": keypool entry invalid");
         }
 
+        m_reserved_key_to_index[nIndex] = keypool.vchPubKey.GetID();
```
In commit "Key pool: Change ReturnDestination interface to take address instead of key" (06d411488c1c669f00a1c947416267941dbb7f7b)

`assert(m_reserved_key_to_index.count(nIndex) == 0);` could be reassuring here

achow101 replied:
Done

Sjors replied:
You lost the assert

achow101 replied:
oops fixed

ryanofsky commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -246,6 +246,8 @@ class LegacyScriptPubKeyMan : public ScriptPubKeyMan, public FillableSigningProv
     int64_t m_max_keypool_index GUARDED_BY(cs_wallet) = 0;
     std::map<CKeyID, int64_t> m_pool_key_to_index;
+    // Tracks keypool indexes to CKeyIDs of keys that have been taken out of the keypool but may be returned to it
+    std::map<int64_t, CKeyID> m_reserved_key_to_index;
```
In commit "Key pool: Change ReturnDestination interface to take address instead of key" (06d411488c1c669f00a1c947416267941dbb7f7b)

Should there be separate map for internal / external key indexes? Or should map key be an (index_number, is_internal) pair? Otherwise how do you know the internal/external index numbers don't overlap?

achow101 replied:
The indexes do not overlap. See https://github.com/bitcoin/bitcoin/blob/master/src/wallet/scriptpubkeyman.cpp#L1083

instagibbs replied:
as noted a few days ago, this should be renamed

achow101 replied:
Renamed it.

ryanofsky commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -281,9 +281,10 @@ void LegacyScriptPubKeyMan::KeepDestination(int64_t index)
 {
-    ReturnKey(index, internal, pubkey);
+    ReturnKey(index, internal, m_reserved_key_to_index[index]);
+    m_reserved_key_to_index.erase(index);
```
In commit "Key pool: Change ReturnDestination interface to take address instead of key" (06d411488c1c669f00a1c947416267941dbb7f7b)

I still don't understand why there is this distinction between ReturnKey and ReturnDestination. Why not just rename ReturnKey to ReturnDestination and get rid of the awkward wrapping?

Previous: https://github.com/bitcoin/bitcoin/pull/17304#discussion_r340311958

achow101 replied:
The wrapper was for the original PR to make it clearer that things were being moved. As that no longer matters, I've added a commit to rename them.

ryanofsky commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -281,9 +281,10 @@ void LegacyScriptPubKeyMan::KeepDestination(int64_t index)
     KeepKey(index);
```
In commit "Key pool: Change ReturnDestination interface to take address instead of key" (06d411488c1c669f00a1c947416267941dbb7f7b)

Why doesn't this need to call m_reserved_key_to_index.erase?

achow101 replied:
It should.

This wasn't necessarily a bug as it doesn't effect behavior.

ryanofsky commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1185,6 +1186,7 @@ bool LegacyScriptPubKeyMan::ReserveKeyFromKeyPool(int64_t& nIndex, CKeyPool& key
             throw std::runtime_error(std::string(__func__) + ": keypool entry invalid");
         }
 
+        m_reserved_key_to_index[nIndex] = keypool.vchPubKey.GetID();
```
In commit "Key pool: Change ReturnDestination interface to take address instead of key" (06d411488c1c669f00a1c947416267941dbb7f7b)

This commit needs a better description. The current handwavy placeholder text doesn't explain very much. Suggest replacing it with the explanation in your PR description "In order for ScriptPubKeyMan to be generic..."

achow101 replied:
Updated the commit message.

ryanofsky reviewed (commented):
I'm almost there...

- [X] b118878bb174941ab6fda8b137166f4c19d0833b Key pool: Move CanGetAddresses call (1/6)
- [X] 863655ae1c1325d9dbb3b1e73ec82f4fe16e9425 Key pool: Move LearnRelated and GetDestination calls (2/6)
- [X] a175268f0e406336cc5d97a0ed458b8f4a805b91 Key pool: Move TopUp call (3/6)
- [ ] 06d411488c1c669f00a1c947416267941dbb7f7b Key pool: Change ReturnDestination interface to take address instead of key (4/6)
- [X] b12121d52f46ac7e8455479bc9ded100cfccc841 Key pool: Make TopUp fail if unexpected wallet flags are set (5/6)
- [X] eb9c2601377d40c4d55798a5ffcbe8c4791b91e5 Key pool: Fix omitted pre-split count in GetKeyPoolSize (6/6)

achow101 committed:
Rename Keep/ReturnKey to Keep/ReturnDestination and remove the wrapper

There is no reason to have Keep/ReturnDestination to be a wrapper for
Keep/ReturnKey. Instead just make them the same function.

Sjors commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1088,15 +1088,16 @@ void LegacyScriptPubKeyMan::AddKeypoolPubkeyWithDB(const CPubKey& pubkey, const
 }
 
-void LegacyScriptPubKeyMan::KeepDestination(int64_t index)
+void LegacyScriptPubKeyMan::KeepDestination(int64_t nIndex)
```
Nit: in the previous commit you renamed `nIndex` to `index`.

achow101 replied:
Fixed

Sjors reviewed (commented):
re-ACK 69533f2 modulo the lost `assert`.

It improves documentation, drops the flag checks in b12121d52f, drops `KeepKey` and `ReturnKey`, adds `erase(nIndex)` to `KeepDestination` (which doesn't change behavior), moves `TopUp()` call under `nIndex == -1`.

The inline discussions have become a bit of a mess, but I believe everything is addressed. @ryanofsky?

ryanofsky reviewed (approved):
Conditional code review ACK 69533f2193556aea7a46b704b6a961dc11957701 if compile error in intermediate commit 1fecf5d0e681a266ea6adb6ae3c695b404cd3b94 is fixed. I also left some other suggestions, which I think would make the changes here safer and easier to understand, and nice to implement before merging.

This code has been pretty difficult to work with and I think every change here makes a solid improvement.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -3252,9 +3250,10 @@ bool ReserveDestination::GetReservedDestination(const OutputType type, CTxDestin
-
     if (nIndex == -1)
     {
+        m_spk_man->TopUp();
```
I've dropped this commit for now.

achow101 replied:
Done as #17537

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1098,15 +1088,15 @@ void LegacyScriptPubKeyMan::AddKeypoolPubkeyWithDB(const CPubKey& pubkey, const
 }
 
-void LegacyScriptPubKeyMan::KeepKey(int64_t nIndex)
+void LegacyScriptPubKeyMan::KeepDestination(int64_t index)
```
Fixed

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1108,7 +1109,9 @@ void LegacyScriptPubKeyMan::ReturnDestination(int64_t index, bool internal, cons
             setExternalKeyPool.insert(nIndex);
         }
-        m_pool_key_to_index[pubkey.GetID()] = nIndex;
+        CKeyID& pubkey_id = m_reserved_key_to_index[nIndex];
```
Done

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1088,15 +1088,16 @@ void LegacyScriptPubKeyMan::AddKeypoolPubkeyWithDB(const CPubKey& pubkey, const
 }
 
-void LegacyScriptPubKeyMan::ReturnDestination(int64_t index, bool internal, const CPubKey& pubkey)
+void LegacyScriptPubKeyMan::ReturnDestination(int64_t nIndex, bool fInternal, const CTxDestination& addr)
```
Done

ryanofsky reviewed (approved):
Code review ACK 42ddfe9125fce4da6738307ef3acf100daac9d75. Changes since last review: just small suggested tweaks and dropping the topup commit

ryanofsky commented:
@instagibbs and @meshcollider, you may be interested to review this. This is code you previously acked when it was part of #16341

instagibbs commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1108,7 +1109,9 @@ void LegacyScriptPubKeyMan::ReturnDestination(int64_t nIndex, bool fInternal, co
             setExternalKeyPool.insert(nIndex);
         }
-        m_pool_key_to_index[pubkey.GetID()] = nIndex;
+        CKeyID& pubkey_id = m_reserved_key_to_index.at(nIndex);
```
Do we really want to crash if that entry doesn't exist?

I guess this is supposed to never happen, since `nIndex==-1` unless it was set otherwise in `GetReservedDestination` where `m_reserved_key_to_index` is also populated.

Sjors commented:
re-ACK 42ddfe9125fce4da6738307ef3acf100daac9d75

achow101 commented:
Rebased.

To resolve some issues during the rebase, I had to change a LearnRelatedScripts into a LearnAllRelatedScripts. This does not change behavior since all of the scripts were implicitly learned anyways.

Sjors commented:
re-ACK 5daefe31c34e12d2a0678aa93e74e94b76d834f5

You could stick to `LearnRelatedScripts` with https://github.com/Sjors/bitcoin/commit/8b75834396b1731f36f57712ba962074b3f7c6c1 but that's not pretty.  The only way `LearnAllRelatedScripts` produces a different result from `LearnRelatedScripts` is if you reserved a legacy key for a legacy change addresses. I'm a bit confused about the other code paths that could end up calling `Learn(All)RelatedScripts`, but even if we don't already, I don't see any harm in adding redundant segwit scripts for legacy change addresses.

promag reviewed (commented):
Code review ACK 5daefe31c34e12d2a0678aa93e74e94b76d834f5.

ryanofsky reviewed (commented):
Code review meh 5daefe31c34e12d2a0678aa93e74e94b76d834f5. Just rebase since last review, but it's kind of a bad rebase. I left a bunch of suggestions to fix, but if you want to just take my suggestions, I pushed a branch implementing them in https://github.com/ryanofsky/bitcoin/commits/review.17373.6-fix

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1101,6 +1102,9 @@ void LegacyScriptPubKeyMan::KeepKey(int64_t nIndex)
     batch.ErasePool(nIndex);
+    CPubKey pk;
+    assert(GetPubKey(m_reserved_key_to_index.at(nIndex), pk));
+    LearnAllRelatedScripts(pk);
```
I don't see a good way to fix that. i don't really like adding the type as an argument to KeepDestination as it isn't useful in other ScriptPubKeyMans.

This also doesn't actually change anything since those scripts are already in memory as they are added via `ImplicitlyLearnRelatedScripts` which is called during key loading. This just writes them to the wallet file.

ryanofsky replied:
> This just writes them to the wallet file.

Wow, that is even worse. If you think this is a good idea please add a comment here explaining that this is doing something nonsensical because design considerations make it not possible for the `KeepDestination` function to be aware of the destination type.

But I'd encourage you to take a look at my straight refactoring of this code in https://github.com/ryanofsky/bitcoin/commits/review.17373.6-fix which doesn't change behavior, doesn't write garbage to files with no explanation, but does (yes) add a `type` context parameter to the `KeepDestination` function, exactly analogous to the `internal` context parameter passed to `ReturnDestination` function, with the understanding that the whole point of having a ScriptPubKeyMan virtual class is to have subclasses that will be implemented differently, and not every implementation will require the exact same information do to every operation.

achow101 replied:
I don't see how this is at all "nonsensical" or "even worse'. But whatever, I've learned my lesson from the multiple days spent stuck on `TopUp()` so it's changed.

achow101 committed:
Add OutputType and CPubKey parameters to KeepDestination

These need to be added so that LearnRelatedScripts can be called
from within KeepDestination later.

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1101,6 +1102,9 @@ void LegacyScriptPubKeyMan::KeepKey(int64_t nIndex)
     WalletBatch batch(m_storage.GetDatabase());
     batch.ErasePool(nIndex);
+    CPubKey pk;
+    assert(GetPubKey(m_reserved_key_to_index.at(nIndex), pk));
```
Done

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1101,6 +1102,9 @@ void LegacyScriptPubKeyMan::KeepKey(int64_t nIndex)
     WalletBatch batch(m_storage.GetDatabase());
     batch.ErasePool(nIndex);
+    CPubKey pk;
+    assert(GetPubKey(m_reserved_key_to_index.at(nIndex), pk));
```
Fixed

ryanofsky commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1091,6 +1092,7 @@ void LegacyScriptPubKeyMan::KeepDestination(int64_t nIndex, const OutputType& ty
     // Remove from key pool
     WalletBatch batch(m_storage.GetDatabase());
     batch.ErasePool(nIndex);
+    LearnRelatedScripts(pubkey, type);
```
In commit "Key pool: Move LearnRelated and GetDestination calls" (094f0545d2d1309ad85eb66f107dec52746490df)

Commit description "There is a minor change in behavior" seems out of date. Would just say "This commit does not change behavior".

The previous description was referring to LearnRelatedScripts call happening moving under the GetReservedDestination `nIndex == -1` check in the old version of this PR before #17237.

achow101 replied:
Fixed.

ryanofsky replied:
Note: This change was done in the wrong commit "Change ReturnDestination interface" (82dbc9344c33217e01c16b27d1418f042d0d0d77) instead the suggested "Move LearnRelated and GetDestination" (393315354da454eb0f9322b92ad736bd23606f97), so `dest` return value is temporarily left unset by the earlier commit and then fixed again in a later unrelated commit. Would be nice to fix but not important.

ryanofsky commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1087,16 +1087,19 @@ void LegacyScriptPubKeyMan::AddKeypoolPubkeyWithDB(const CPubKey& pubkey, const
 }
 
-void LegacyScriptPubKeyMan::KeepDestination(int64_t nIndex, const OutputType& type, const CPubKey& pubkey)
+void LegacyScriptPubKeyMan::KeepDestination(int64_t nIndex, const OutputType& type)
```
In commit "Key pool: Change ReturnDestination interface to take address instead of key" (8e70f2b4150ca2d25f06cc65295cb27a86c49681)

This should call `m_reserved_key_to_index.erase(nIndex);` to clean up

achow101 replied:
Fixed.

ryanofsky reviewed (approved):
Code review ACK 562a4b0e1c5362689ff63f744737b93abd88677c, but erase bug should probably be fixed.

re: https://github.com/bitcoin/bitcoin/pull/17373#discussion_r350870468

> I don't see how this is at all "nonsensical" or "even worse'. But whatever, I've learned my lesson from the multiple days spent stuck on TopUp() so it's changed

Sorry if I am being too stringent in reviewing, or if I'm just failing to understand things that are more obvious to you. I am trying to be helpful and push these changes along, but occasionally I do wind up strong opinions about minor things and try to express them without getting in the way. In this case, I implemented all the suggestions I made for you, so I wasn't trying to waste your time or make you do extra work.

Writing data that will never be used is nonsensical because it doesn't make sense. Writing garbage to disk is "even worse" than wasting memory because a clunky runtime behavior can go away with a simple code change while junk in the database can stick around forever and cause difficult to reproduce problems in the future.

instagibbs reviewed (changes requested):
`m_reserved_key_to_index` should be renamed(it's backwards), and agree with @ryanofsky suggested fixes/cleanups

contingent utACK https://github.com/bitcoin/bitcoin/pull/17373/commits/562a4b0e1c5362689ff63f744737b93abd88677c

instagibbs commented:
utACK https://github.com/bitcoin/bitcoin/pull/17373/commits/66b82896d5c9d40abb9e62a7035a97025aa8aad3

only changes are the ones suggested in my previous review

ryanofsky reviewed (approved):
Code review ACK 66b82896d5c9d40abb9e62a7035a97025aa8aad3. Just suggested changes since last review: updating commit message, renaming map and fixing erase

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -3305,31 +3305,27 @@ bool ReserveDestination::GetReservedDestination(CTxDestination& dest, bool inter
         fInternal = keypool.fInternal;
         address = dest;
     }
-    assert(vchPubKey.IsValid());
```
Note to other reviewers: this assert is lost, but `GetReservedDestination` calls `ReserveKeyFromKeyPool` which throws if `!keypool.vchPubKey.IsValid()`

Sjors reviewed (approved):
re-ACK 66b8289

Nit: 5a8a2f20b4408ae71750a3254a4b74cc6d5b1bf3 description of still uses `m_reserved_key_to_index` instead of `m_reserved_key_to_index`

@ryanofsky wrote:
> Sorry if I am being too stringent in reviewing, or if I'm just failing to understand things that are more obvious to you. 

The bus factor (and test coverage) of our wallet code is still uncomfortably low, and there's a reason we're sticking legacy code in a box rather than completely refactoring it. Wallet bugs ain't fun. It's worth some extra effort to be careful and/or clearly explain stuff.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -3302,23 +3302,21 @@ bool ReserveDestination::GetReservedDestination(CTxDestination& dest, bool inter
         vchPubKey = keypool.vchPubKey;
         fInternal = keypool.fInternal;
+        address = dest;
     }
```
In commit "Key pool: Move LearnRelated and GetDestination calls" (393315354da454eb0f9322b92ad736bd23606f97

Do we need:

```diff
@@ -3308,6 +3308,8 @@ bool ReserveDestination::GetReservedDestination(CTxDestination& dest, bool inter
         vchPubKey = keypool.vchPubKey;
         fInternal = keypool.fInternal;
         address = dest;
+    } else {
+        dest = address;
     }
     assert(vchPubKey.IsValid());
     return true;
```

It seems like `dest` is currently left unset in the `nIndex != -1` case where previously it was always set.

achow101 replied:
Done as suggested.

achow101 replied:
Oops, moved it to the right commit.

ryanofsky reviewed (approved):
Code review ACK 244f5c23f864068242e6b4721577aaa470762ed5. Just suggested dest return fix since the last review. Thanks for the fix.

instagibbs commented on src/wallet/wallet.cpp:
```diff
@@ -3298,43 +3298,34 @@ bool ReserveDestination::GetReservedDestination(CTxDestination& dest, bool inter
     }
-    assert(vchPubKey.IsValid());
-    address = GetDestinationForKey(vchPubKey, type);
     dest = address;
```
future work: address vs dest clarification via renaming would be great, even just `m_dest` vs `dest`

instagibbs reviewed (commented):
code review re-ACK https://github.com/bitcoin/bitcoin/pull/17373/commits/886f1731bec4393dd342403ac34069a3a4f95eea

ryanofsky reviewed (approved):
Code review ACK 886f1731bec4393dd342403ac34069a3a4f95eea. Only change is moving earlier fix to a better commit (same end result).

Sjors commented:
Code review re-ACK 886f1731bec4393dd342403ac34069a3a4f95eea

(original nit about `m_reserved_key_to_index` in 386a994b853bc5b3a2ed0d812673465b8ffa4849 description still stands, not worth losing ACKs)

promag commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1112,13 +1112,15 @@ void LegacyScriptPubKeyMan::ReturnKey(int64_t nIndex, bool fInternal, const CPub
-        m_pool_key_to_index[pubkey.GetID()] = nIndex;
+        CKeyID& pubkey_id = m_index_to_reserved_key.at(nIndex);
+        m_pool_key_to_index[pubkey_id] = nIndex;
+        m_index_to_reserved_key.erase(nIndex);
```
nit, could avoid 2nd lookup:
```
auto it = find(index)
...
erase(it)
```

achow101 replied:
Leaving as is to keep ACKs

promag commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1179,6 +1181,8 @@ bool LegacyScriptPubKeyMan::ReserveKeyFromKeyPool(int64_t& nIndex, CKeyPool& key
         }
 
+        assert(m_index_to_reserved_key.count(nIndex) == 0);
+        m_index_to_reserved_key[nIndex] = keypool.vchPubKey.GetID();
```
nit, could also avoid 2nd lookup:
```
auto res = emplace(nIndex, ...)
assert(res.second)
```

achow101 replied:
Leaving as is to keep ACKs

promag reviewed (commented):
Code review ACK 886f1731bec4393dd342403ac34069a3a4f95eea.
